src/proto_reader.rs

```diff
@@ -0,0 +1,169 @@
+use std::collections::{HashMap, VecDeque};
+
+use anyhow::Context;
+use futures_util::future::join_all;
+use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};
+use protox::file::{FileResolver, GoogleFileResolver};
+
+use crate::resource_reader::ResourceReader;
+use crate::runtime::TargetRuntime;
+
+pub struct ProtoReader {
+    resource_reader: ResourceReader,
+}
+
+pub struct ProtoMetadata {
+    pub package: Option<String>,
+    pub descriptor_set: FileDescriptorSet,
+}
+
+impl ProtoReader {
+    pub fn init(runtime: TargetRuntime) -> Self {
+        Self { resource_reader: ResourceReader::init(runtime) }
+    }
+
+    pub async fn resolve_protos<T: AsRef<str>>(
```

Comment:
doesn't need to be public.

Reply:
Should only have `read` and `read_all` methods as public.

Reply:
this is the new function I introduced to resolve all the protos.. we can directly use this in config reader and generator reader without having to worry about parent and resolved descriptors